lib/recursive-methods.rb

```diff
@@ -1,49 +1,71 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    raise ArgumentError.new("can't accept negative numbers") if n < 0
+    return 1 if n == 0
+    return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2)
+# Space complexity: O(n^2)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    return s if s.length <= 1 # base case
+    return s[-1] + reverse(s[1..-2]) + s[0] #recursive case
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2)
+# Space complexity: O(n^2)
 def reverse_inplace(s)
```

Comment:
⚠️ This is not reversed in place.  You're creating a new string with each recursive call.  

Think about how to reverse it with this method signature.

`def reverse_inplace(s, low = 0, high = s.length - 1)`